src/Microsoft.OData.Client/ClientEdmModel.cs

```diff
@@ -591,32 +591,60 @@ private IEdmProperty CreateEdmProperty(IEdmStructuredType declaringType, Propert
                 propertyEdmType.TypeKind == EdmTypeKind.Collection,
                 "Property kind should be Entity, Complex, Enum, Primitive or Collection.");
 
-            IEdmProperty edmProperty;
+            IEdmProperty edmProperty = null;
             bool isPropertyNullable = ClientTypeUtil.CanAssignNull(propertyInfo.PropertyType);
             if (propertyEdmType.TypeKind == EdmTypeKind.Entity || (propertyEdmType.TypeKind == EdmTypeKind.Collection && ((IEdmCollectionType)propertyEdmType).ElementType.TypeKind() == EdmTypeKind.Entity))
             {
-                IEdmEntityType declaringEntityType = declaringType as IEdmEntityType;
-                if (declaringEntityType == null)
+                if (declaringType.TypeKind == EdmTypeKind.Entity)
                 {
-                    throw c.Error.InvalidOperation(c.Strings.ClientTypeCache_NonEntityTypeCannotContainEntityProperties(propertyInfo.Name, propertyInfo.DeclaringType.ToString()));
+                    IEdmEntityType declaringEntityType = declaringType as IEdmEntityType;
+                    if (declaringEntityType == null)
+                    {
+                        throw c.Error.InvalidOperation(c.Strings.ClientTypeCache_NonEntityTypeCannotContainEntityProperties(propertyInfo.Name, propertyInfo.DeclaringType.ToString()));
+                    }
+
+                    // Create a navigation property representing one side of an association.
+                    // The partner representing the other side exists only inside this property and is not added to the target entity type,
+                    // so it should not cause any name collisions.
+                    edmProperty = EdmNavigationProperty.CreateNavigationPropertyWithPartner(
+                        ClientTypeUtil.GetServerDefinedName(propertyInfo),
+                        propertyEdmType.ToEdmTypeReference(isPropertyNullable),
+                        /*dependentProperties*/ null,
+                        /*principalProperties*/ null,
+                        /*containsTarget*/ false,
+                        EdmOnDeleteAction.None,
+                        "Partner",
+                        declaringEntityType.ToEdmTypeReference(true),

```

Comment:
This seems the only difference between this block and the else block below... Could have a test to create it for both complex and entity types and if the containing type is none of the two then throw an exception with a more correct(?) message...